### PR TITLE
__array_ufunc__ *args and 'out' always a tuple

### PR DESCRIPTION
**Pull request by mhvk**
(comments in main thread)

### CODE REVIEW

File: doc/source/reference/arrays.classes.rst [hunk 1]
```diff
@@ -39,27 +39,26 @@ Special attributes and methods
...
+   similarly to Python's ``__mul__`` and other binary operation routines.
 
    - *ufunc* is the ufunc object that was called.
    - *method* is a string indicating which Ufunc method was called
      (one of ``"__call__"``, ``"reduce"``, ``"reduceat"``,
      ``"accumulate"``, ``"outer"``, ``"inner"``).
-   - *inputs* is a tuple of the input arguments to the ``ufunc``
+   - *inputs* is a tuple of the input arguments to the ``ufunc``.
    - *kwargs* is a dictionary containing the optional input arguments
      of the ufunc. The ``out`` argument is always contained in
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: ```All  ``out`` arguments, positional or keyword, are contained in ... as a tuple if given.```

File: doc/source/reference/arrays.classes.rst [hunk 2]
```diff
@@ -73,40 +72,26 @@ NumPy provides several hooks that classes can customize:
    If an :class:`ndarray` subclass defines the :func:`__array_ufunc__`
    method, this disables the :func:`__array_wrap__`,
    :func:`__array_prepare__`, :data:`__array_priority__` mechanism
-   described below.
+   described below (which may eventually be deprecated).  Furthermore, it
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: ```and which may eventually be deprecated```?

Comment (charris) [inline doc/source/reference/arrays.classes.rst]: ```Furthermore, unless it is set to  `None`,```

File: doc/source/reference/arrays.classes.rst [hunk 3]
```diff
@@ -73,40 +72,26 @@ NumPy provides several hooks that classes can customize:
...
+   takes precedence over right-hand binary operator override methods (such
+   as ``__rmul__``) or comparison operations (such as ``__gt__``) in your
+   class: the presumption is that your class will already be called first
+   via :func:`__array_ufunc__`.
+
+   Like with other methods in python, such as ``__hash__`` and
+   ``__iter__``, it is possible to indicate that your class does *not*
+   support ufuncs by setting ``__array_ufunc__ = None`` (assuming your
+   class is not an :class:`ndarray` subclass). With this, inside ufuncs,
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: Remove `(` and `if you class...`.

Does this imply the subclasses cannot opt out?

Comment (charris) [inline doc/source/reference/arrays.classes.rst]: ```With this ufuncs will treat your class..```

Comment (charris) [inline doc/source/reference/arrays.classes.rst]: Hmm, I think it needs to be stated that that for the numeric binary ops this is equivalent to returning NotImplemented.

File: doc/source/reference/arrays.classes.rst [hunk 4]
```diff
@@ -73,40 +72,26 @@ NumPy provides several hooks that classes can customize:
...
+   your class will be treated as if it returned :obj:`NotImplemented`
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: I don't think this is correct. It is the calling class, ndarray and subclasses thereof, for which it is equivalent to calling `NotImplemented`. The effect is that unary ops fail and binary ops get passed to the corresponding right op. Or am I missing something?

File: doc/source/reference/arrays.classes.rst (hunk 3 above)
Comment (mhvk) [inline doc/source/reference/arrays.classes.rst]: Good comments; I'll implement them later.

For now, in answer to your question, I think the text is not quite right: for a proper subclass of ndarray, the right operators will be given preference by CPython already.

File: doc/source/reference/arrays.classes.rst (hunk 4 above)
Comment (mhvk) [inline doc/source/reference/arrays.classes.rst]: I'll try to separate more clearly what happens for binary operators and inside ufuncs (I've gotten myself confused right now...).
p.s.  Will do this over the weekend -- some horrible money proposal is taking all my time right now.

Comment (charris) [inline doc/source/reference/arrays.classes.rst]: No problem. We all need to earn a living...

**Comment by mhvk** [comment]
@charris - rebased, and now with updated documentation (which I think may correspond to reality...)

File: doc/source/reference/arrays.classes.rst [hunk 5]
```diff
@@ -39,74 +39,70 @@ Special attributes and methods
...
-.. method:: class.__array_ufunc__(ufunc, method, inputs, **kwargs)
+.. method:: class.__array_ufunc__(ufunc, method, *inputs, **kwargs)
 
-   .. versionadded:: 1.11
+   .. versionadded:: 1.13
 
-   Any class (ndarray subclass or not) can define this method to
-   override behavior of NumPy's ufuncs. This works quite similarly to
-   Python's ``__mul__`` and other binary operation routines.
+   Any class (ndarray subclass or not) can define this method (or set it to
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: Commas, not "(ndarray....)". I'd also remove the parenthesis around "or set...".

Comment (charris) [inline doc/source/reference/arrays.classes.rst]: I'll do the correction.

File: doc/source/reference/arrays.classes.rst [hunk 6]
```diff
@@ -39,74 +39,70 @@ Special attributes and methods
...
+   methods such as ``__radd__``, ``__le__``, etc.). If
+   ``__array_ufunc__ = None``, however, forward methods on
+   :class:`ndarray` will unconditionally return :obj:`NotImplemented`,
+   so that your reverse methods will get called.
+
+   .. note:: If you subclass :class:`ndarray`, we strongly recommend
+      that you avoid confusion by neither setting
+      :func:`__array_ufunc__` to :obj:`None` (which seems to make no
+      sense for an array subclass), nor defining it and also defining
+      reverse methods (which will be called by ``CPython`` in
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: Remove parenthesis.

File: doc/source/reference/arrays.classes.rst [hunk 7]
```diff
@@ -129,6 +125,9 @@ NumPy provides several hooks that classes can customize:
    the subclass and update metadata before returning the array to the
    ufunc for computation.
 
+   .. note:: It is hoped to eventually deprecate this method in favour of
+             :func:__array_ufunc__`.
```
Comment (charris) [inline doc/source/reference/arrays.classes.rst]: Missing backtick.

File: numpy/core/src/private/ufunc_override.c [hunk 8]
```diff
@@ -46,12 +50,13 @@ normalize_reduce_args(PyUFuncObject *ufunc, PyObject *args,
...
+    *normal_args = PyTuple_GetSlice(args, 0, 1);
+    for (i = 1; i < nargs; i++) {
         obj = PyTuple_GET_ITEM(args, i);
-        if (i == 0) {
-            *normal_args = PyTuple_GetSlice(args, 0, 1);
+        if (obj == Py_None) {
+            continue;
         }
-        else if (i == 1) {
+        if (i == 1) {
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: We could eventually turn these into a switch statement.

Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Could even make a switch on nargs and complete by falling through in reverse order, thus avoiding the loop.

File: numpy/core/src/private/ufunc_override.c [hunk 9]
```diff
@@ -263,6 +273,30 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
...
+                    all_none = (PyTuple_GET_ITEM(out, i) == Py_None);
+                    if (!all_none) {
+                        break;
+                    }
+                }
+                if(all_none){
+                    PyDict_DelItemString(normal_kwds, "out");
+                }
+            }
+            else if (out != Py_None) {  /* not already a tuple and not None */
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Put trailing comment on next line.

File: numpy/core/src/private/ufunc_override.c [hunk 10]
```diff
@@ -263,6 +273,30 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
     /* Build new kwds */
     if (kwds && PyDict_CheckExact(kwds)) {
         normal_kwds = PyDict_Copy(kwds);
+        /* ensure out is always a tuple */
+        PyObject *out = PyDict_GetItemString(normal_kwds, "out");
+        if (out != NULL) {
+            if(PyTuple_Check(out)) {
+                int all_none;
+                int i;
+                for (i==0; i<PyTuple_GET_SIZE(out); i++) {
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Bank line after declarations.

File: numpy/core/src/private/ufunc_override.c [hunk 11]
```diff
@@ -394,13 +435,13 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
...
-    Py_XDECREF(normal_args);
     Py_XDECREF(normal_kwds);
+    Py_DECREF(override_args);
     return 0;
 
 fail:
     Py_XDECREF(method_name);
-    Py_XDECREF(normal_args);
     Py_XDECREF(normal_kwds);
+    Py_DECREF(override_args);
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: This is safe?

File: numpy/core/src/private/ufunc_override.c [hunk 12]
```diff
@@ -306,12 +340,26 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
         goto fail;
     }
 
+    Py_ssize_t len = PyTuple_GET_SIZE(normal_args);
+    PyObject *override_args = PyTuple_New(len + 2);
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Really need to check for NULL here. Then you will want Py_XDECREF in failure path.

File: numpy/core/src/private/ufunc_override.c [hunk 13]
```diff
@@ -263,6 +273,30 @@ PyUFunc_CheckOverride(PyUFuncObject *ufunc, char *method,
...
+                    if (!all_none) {
+                        break;
+                    }
+                }
+                if(all_none){
+                    PyDict_DelItemString(normal_kwds, "out");
+                }
+            }
+            else if (out != Py_None) {  /* not already a tuple and not None */
+                PyObject *out_tuple = PyTuple_New(1);
```
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Need error check here.

File: numpy/core/src/private/ufunc_override.c (hunk 11 above)
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: No, it isn't if an error check is performed.

**Comment by charris** [comment]
Do you want to correct the comments? If you are short on time I can do it after the merge.

**Comment by charris** [comment]
Note that the normalization routines still need error checking at some point.

**Comment by charris** [comment]
Merged, I'll do a maintenance PR on top of this.

**Comment by mhvk** [comment]
OK, thanks! I'll have a look at those just in case, and then rebase #5 on top.

File: numpy/core/src/private/ufunc_override.c (hunk 10 above)
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Also `i==0` and spaces around operators.

File: numpy/core/src/private/ufunc_override.c (hunk 12 above)
Comment (charris) [inline numpy/core/src/private/ufunc_override.c]: Also, no declarations except at top of blocks.

**Comment by charris** [comment]
I merged a fixup with this as it doesn't compile as is.

**Comment by mhvk** [comment]
Thanks for all the corrections; I'm amazed at how much I missed (and that it compiled...)

**Pull request merged**